Add Cloud Run Jobs Infrastructure Block

peytonrunyan opened the pull request:
<!-- Thanks for contributing to prefect-gcp! 🎉-->

## Summary
<!-- A brief summary explaining the purpose of this PR -->
Adds a `CloudRunJob` infrastructure block to allow for running deployments using Google Cloud Run Jobs. 
## Relevant Issue(s)
<!-- If this PR addresses any open issues, please let us know which one here -->
Closes https://github.com/PrefectHQ/prefect/issues/5884

## Checklist
- [x] Summarized PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-gcp/blob/main/CHANGELOG.md)
- [x] Finished unit tests
- [ ] Incorporated Michael's work on https://github.com/PrefectHQ/prefect/pull/6622
- [ ] Figured out how to make this prettier on the UI
- [x] Added tests to Credentials changes

## Details and discussion
I want to go ahead and get this under review while I'm finishing up the last of the tests for what I have. I plan to schedule a walk-through for either this afternoon or tomorrow afternoon if possible. 

**Credentials**
I made some changes to the Credentials block that I would like to discuss with @ahuang11 before anything is finalized. 
- I moved some of the validation out of the function body to separate validators 
- I made `get_service_account_value` public so that I could access it from within `CloudRunJob`
- I removed project as a field and made it a property because it is contained within the service account info, and seems like an extra point of failure if the provided project doesn't match the service account
- I also added a method getting an access token, which is no longer being used in my code, but will probably be useful for future work.

**Design Decisions**
- A unique cloud run job can only have one instance running. After some discussion with @madkinsz and @anna-geller we decided to just create a new job each time, instead of checking to see if the job is free and creating a new job if not.
- Cloud Run Jobs can only pull from google cloud registries, so no default image option is available
- Resource requests (cpu and memory) cannot be higher than limits, so when specifying a resource amount, both the limit and the request are set to the same number

**Open Question and things I'm kicking around**
- I have a couple of error response formatting functions that aren't especially well fleshed out. 
- The client calls could possibly be moved to `Job` and `Execution` and out of `CloudRunJob`. This would probably make testing easier. This feels like it could probably be done later.

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+import json
+import re
+import time
+from ast import Delete
```
Is this used?

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+from typing import Literal, Optional
+from uuid import uuid4
+
+import googleapiclient
```
Do we need to add this to `requirements.txt`

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+    def has_execution_in_progress(self) -> bool:
+        """See if job has a run in progress."""
+        return (
+            self.execution_status == {}
```
```suggestion
            not self.execution_status
```

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+
+    Project name information is provided by the Credentials object, and should always
+    be correct as long as the Credentials object is for the correct project.
+    """
```
Should there be docstrings of fields?

desertaxle replied:
`mkdocstrings` should display the pydantic fields for us.

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+    credentials: GcpCredentials
+
+    # Job settings
+    cpu: Optional[int] = Field(
+        description=(
```
Do you need  to explicitly specify `default=None` like `Field(default=None, ...)`

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+        ),
+    )
+
+    region: str
+    credentials: GcpCredentials
```
Missing field descriptions

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+            "based on the rules specified at https://cloud.google.com/run/docs/configuring/memory-limits#setting-jobs ."
+        )
+    )
+    args: Optional[list[str]] = Field(
```
```suggestion
    args: Optional[List[str]] = Field(
```

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+            "Arguments to be passed to your Cloud Run Job's entrypoint command."
+        )
+    )
+    env: dict[str, str] = Field(
```
```suggestion
    env: Dict[str, str] = Field(
```

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+            components = self.image.split("/")
+            # gcr.io/<project_name>/repo/whatever
+            image_name = components[2]
+            modified_image_name = image_name.replace((":"), "-").replace(
+                ("."), "-"
```
Are the extra pairs of parentheses necessary?
```suggestion
            modified_image_name = image_name.replace(":", "-").replace(
                ".", "-"
```

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+            if len(name) > 50:
+                self.logger.warning(
+                    "Job names must be 50 or fewer characters in length. Shortening "
+                    f"job name from {name} to {name[:50]}."
```
https://peps.python.org/pep-3101/
```suggestion
                    f"job name from {name!r} to {name[:50]!r}."
```

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+        """Make sure memory conforms to expected values for API.
+        See: https://cloud.google.com/run/docs/configuring/memory-limits#setting
+        """
+        _, size, unit = re.split("(\d+)", value)
```
What if people provide a unit of ` MiB` or `GiB`?

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+
+        return job_execution
+
+    def _wait_for_job_creation(self, client: Resource, poll_interval: int = 5):
```
Should this be async?

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+    async def run(self, task_status: Optional[TaskStatus] = None):
+        with self._get_jobs_client() as jobs_client:
+            try:
+                self.logger.info(f"Creating Cloud Run Job {self.job_name}")
```
Maybe move this outside the try

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+                command = (
+                    " ".join(self.command)
+                    if self.command
+                    else "'default container command'"
```
The command would have double `''default container command''`

```suggestion
                    else "default container command"
```

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+                )
+
+                self.logger.info(
+                    f"Cloud Run Job {self.job_name}: Running command '{command}'"
```
```suggestion
                    f"Cloud Run Job {self.job_name!r}: Running command {command!r}"
```

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+        with self._get_jobs_client() as jobs_client:
+            try:
+                self.logger.info(f"Creating Cloud Run Job {self.job_name}")
+                self._create_job(jobs_client=jobs_client, body=self._jobs_body())
+            except googleapiclient.errors.HttpError as exc:
+                self._create_job_error(exc)
```
Would these need to be async?

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+            )
+        except Exception as exc:
+            self.logger.exception(
+                f"Received an unexpected exception while monitoring Cloud Run Job '{self.job_name}':\n{exc!r}"
```
```suggestion
                f"Received an unexpected exception while monitoring Cloud Run Job {self.job_name!r}:\n{exc!r}"
```

ahuang11 commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,578 @@
+
+        return json.dumps(body, indent=2)
+
+    def _watch_job_execution(self, job_execution: Execution, poll_interval=5):
```
```suggestion
    def _watch_job_execution(self, job_execution: Execution, poll_interval: int = 5):
```

peytonrunyan commented:
@ahuang11 I did a good chunk of refactoring like we discussed in the walk-through and also addressed all of the points that I've marked as resolved. I need to finish up some tests, check out the remaining unresolved points, and then I should be ready for one final review.

ahuang11 commented:
Did some preliminary cleanup with my last commit. Commented out / removed some lines of tests; not entirely sure if tests are outdated, or the code is not doing what it's supposed to be doing.

Will be do more testing soon.

ahuang11 commented:
Not sure why tests pass locally, but encounter this on GitHub
```
tests/test_cloud_run_job.py:11: in <module>
    from prefect_gcp.cloud_run_job import CloudRunJob, CloudRunJobResult
prefect_gcp/__init__.py:3: in <module>
    from .cloud_run_job import gcp_cloud_run_job as cloud_run_job  # noqa
prefect_gcp/cloud_run_job/__init__.py:1: in <module>
    from .gcp_cloud_run_job import CloudRunJob, CloudRunJobResult  # noqa
prefect_gcp/cloud_run_job/gcp_cloud_run_job.py:62: in <module>
    class CloudRunJob(Infrastructure):
prefect_gcp/cloud_run_job/gcp_cloud_run_job.py:160: in CloudRunJob
    def _remove_image_spaces(cls, value):
pydantic/class_validators.py:93: in pydantic.class_validators.validator.dec
    ???
pydantic/class_validators.py:152: in pydantic.class_validators._prepare_validator
    ???
E   pydantic.errors.ConfigError: duplicate validator function "prefect_gcp.cloud_run_job.gcp_cloud_run_job.CloudRunJob._remove_image_spaces"; if this is intended, set `allow_reuse=True`
```

ahuang11 commented:
Wooo tests finally pass XD

ahuang11 commented:
Not entirely sure about:
"""
Incorporated Michael's work on https://github.com/PrefectHQ/prefect/pull/6622
Figured out how to make this prettier on the UI
"""

Also, have to run an actual test tomorrow, but should be good for a round of reviews first

ahuang11 commented:
`cloud_run_job.preview()` shows the prefect api key; not sure if that should be included.
```
{
  "apiVersion": "run.googleapis.com/v1",
  "kind": "Job",
  "metadata": {
    "name": "dummy-66711637f9c14fb9b10fffa0d7ef2c57",
    "annotations": {
      "run.googleapis.com/launch-stage": "BETA"
    }
  },
  "spec": {
    "template": {
      "spec": {
        "template": {
          "spec": {
            "containers": [
              {
                "image": "gcr.io/project/dummy",
                "env": [
                  {
                    "name": "PREFECT_API_URL",
                    "value": "api-url"
                  },
                  {
                    "name": "PREFECT_API_KEY",
                    "value": "pnu_mysecretkey"
                  }
                ],
                "command": [
                  "echo",
                  "hello world"
                ]
              }
            ]
          }
        }
      }
    }
  }
}
```

ahuang11 commented:
Tested `run()` and it successfully ran!
![image](https://user-images.githubusercontent.com/15331990/190223921-7e0e29d6-f68b-43f6-b2ce-06d10b1f6fa0.png)

```
from prefect import flow, task
from prefect_gcp import GcpCredentials
from prefect_gcp.cloud_run_job import CloudRunJob

@flow
def test_flow():
    cloud_run_job = CloudRunJob(
        image="us-docker.pkg.dev/cloudrun/container/job:latest",
        credentials=GcpCredentials.load("infra-block"),
        region="us-central1",
        command=["echo", "hello world"],
    )
    return cloud_run_job.run()

test_flow()
```

ahuang11 commented:
Would be nice if UI had it sorted by "required" then "optional"
![image](https://user-images.githubusercontent.com/15331990/190226601-4c33d502-91e3-435f-95c9-8de48f8e0325.png)

Ran successfully:
```
from prefect import flow, task
from prefect_gcp import GcpCredentials
from prefect_gcp.cloud_run_job import CloudRunJob

@flow
def test_flow():
    cloud_run_job_block = CloudRunJob.load("infra-block-test")
    return cloud_run_job_block.run()

test_flow()
```

```
10:50:14.985 | INFO    | prefect.engine - Created flow run 'logical-jackrabbit' for flow 'test-flow'
/Users/andrew/Applications/python/prefect/src/prefect/blocks/core.py:649: UserWarning: Block document has schema checksum sha256:8fec0d10b2a2092fc52574443cea045d9f5230f4d9cbbb0df7e7a45be9bb38ff which does not match the schema checksum for class 'CloudRunJob'. This indicates the schema has changed and this block may not load.
  return cls._from_block_document(block_document)
10:50:18.894 | INFO    | prefect.infrastructure.cloud-run-job - Creating Cloud Run Job container-a791824ab5de4788b6a8a72ed0e09b69
10:50:19.905 | INFO    | prefect.infrastructure.cloud-run-job - Submitting Cloud Run Job container-a791824ab5de4788b6a8a72ed0e09b69 for execution.
10:50:20.751 | INFO    | prefect.infrastructure.cloud-run-job - Cloud Run Job 'container-a791824ab5de4788b6a8a72ed0e09b69': Running command 'default container command'
10:50:49.240 | INFO    | prefect.infrastructure.cloud-run-job - Job Run container-a791824ab5de4788b6a8a72ed0e09b69 completed successfully
10:50:49.242 | INFO    | prefect.infrastructure.cloud-run-job - Job Run logs can be found on GCP at: 
```

madkinsz commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,678 @@
+    interact with the returned objects.
+    """
+
+    metadata: dict
+    spec: dict
+    status: dict
+    name: str
+    ready_condition: dict
+    execution_status: dict
```
I think we don't need to necessarily go crazy writing strict types for our internal interactions with these complex external services.

madkinsz commented on requirements.txt:
```diff
@@ -1 +1,2 @@
-prefect>=2.0b9
\ No newline at end of file
+prefect>=2.0.0
+google-api-python-client>=2.20.0
```
Directly embedding requirements in `setup.py` isn't a great practice. It does make some sense to match the collection template for now if that's the case though.

desertaxle replied:
We have other libraries in `setup.py` because they're extras. I'm ok with having this library in `requirement.txt` since it's more generalized compared to the service-specific client libraries.

madkinsz commented:
> cloud_run_job.preview() shows the prefect api key; not sure if that should be included.

Should be handled by "Secret" settings rather than manual exclusion

ahuang11 commented:
> > cloud_run_job.preview() shows the prefect api key; not sure if that should be included.
> 
> Should be handled by "Secret" settings rather than manual exclusion

I think we need a SecretJson field type to support service_account_info

desertaxle commented on docs/cloud_run_job.md:
```diff
@@ -0,0 +1 @@
+::: prefect_gcp.cloud_run_job
```
I think the new module should be named `cloud_run` since that's the name of the GCP service.
```suggestion
::: prefect_gcp.cloud_run
```

peytonrunyan replied:
FWIW, Cloud Run has two distinct services - Cloud Run Services and Cloud Run Jobs. This only deals with Cloud run Jobs.
https://cloud.google.com/run/docs/overview/what-is-cloud-run#services-and-jobs

desertaxle replied:
Yep, the block will still be called `CloudRunJob`. I'm only suggesting that we change the name of the module to match the granularity of other modules in this collection. If we develop other integrations that work with Cloud Run services or Cloud Run jobs, they can go into this module.

desertaxle commented on mkdocs.yml:
```diff
@@ -61,3 +61,4 @@ nav:
     - Cloud Storage: cloud_storage.md
     - BigQuery: bigquery.md
     - Secret Manager: secret_manager.md
+    - Cloud Run Job: cloud_run_job.md
```
```suggestion
    - Cloud Run: cloud_run.md
```

desertaxle commented on setup.py:
```diff
@@ -40,6 +40,7 @@
     entry_points={
         "prefect.collections": [
             "prefect_gcp = prefect_gcp",
+            "prefect_gcp_cloud_run_job = prefect_gcp.cloud_run_job",
```
```suggestion
            "prefect_gcp_cloud_run = prefect_gcp.cloud_run",
```

desertaxle commented on prefect_gcp/credentials.py:
```diff
@@ -84,62 +85,107 @@ class GcpCredentials(Block):
+    @property
+    def project_id(self):
+        if self._project_id is None:
+            if self.project:
+                self._project_id = self.project
+            else:
+                self._project_id = (
+                    self.get_credentials_from_service_account().project_id
+                )
```
This initialization probably belongs in the `block_initialization` method

desertaxle commented on prefect_gcp/credentials.py:
```diff
@@ -84,62 +85,107 @@ class GcpCredentials(Block):
+        """  # noqa
+        request = google.auth.transport.requests.Request()
+        credentials = self.get_credentials_from_service_account()
+        credentials.refresh(request)
```
I would guess that this makes a network call, so it should probably be run in a worker thread and this method should be async.

desertaxle commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,686 @@
+            execution_status=cls._get_execution_status(response),
+        )
+
+    @classmethod
```
This doesn't use `cls` and could be a static method
```suggestion
    @staticmethod
```

desertaxle commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,686 @@
+        response = request.execute()
+        return response
+
+    @classmethod
```
This doesn't use `cls` and could be a static method
```suggestion
    @staticmethod
```

desertaxle commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,686 @@
+        response = request.execute()
+        return response
+
+    @classmethod
```
This doesn't use `cls` and could be a static method
```suggestion
    @staticmethod
```

desertaxle commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,686 @@
+                except Exception as exc:
+                    self.logger.exception(
+                        "Received an unexpected exception while attempting to delete"
+                        f" Cloud Run Job.'{self.job_name}':\n{exc!r}"
```
`logger.exception` [includes the exception info by default](https://docs.python.org/3/library/logging.html#logging.exception).
```suggestion
                        f" Cloud Run Job.'{self.job_name}'"
```

desertaxle commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,686 @@
+        except Exception as exc:
+            self.logger.exception(
+                "Received an unexpected exception while monitoring Cloud Run Job"
+                f" '{self.job_name!r}':\n{exc!r}"
```
```suggestion
                f" '{self.job_name!r}'"
```

desertaxle commented on prefect_gcp/cloud_run_job.py:
```diff
@@ -0,0 +1,686 @@
+            except Exception as exc:
+                self.logger.exception(
+                    "Received an unexpected exception while attempting to delete Cloud"
+                    f" Run Job.'{self.job_name}':\n{exc!r}"
```
```suggestion
                    f" Run Job.'{self.job_name}'"
```

ahuang11 commented:
Ran successfully with new changes
```
from prefect import flow, task
from prefect_gcp import GcpCredentials
from prefect_gcp.cloud_run import CloudRunJob

@flow
def test_flow():
    cloud_run_job = CloudRunJob(
        image="us-docker.pkg.dev/cloudrun/container/job:latest",
        credentials=GcpCredentials.load("infra-block"),
        region="us-central1",
        command=["echo", "hello world"],
        project="abc"
        keep_job=True
    )
    return cloud_run_job.run()

test_flow()
```

![image](https://user-images.githubusercontent.com/15331990/191116683-a3c128c1-e6f9-40d2-8247-5dbd5a1d3132.png)

ahuang11 commented:
Thanks @peytonrunyan for contributing the bulk of this PR!

peytonrunyan commented:
Whoop - thanks for taking this to the finish line man!
